### PR TITLE
feat: per-document-access

### PR DESCRIPTION
**Pull request by janl**
This is the first per-doc-access PR worth sharing more widely.

This allows the creation of databases with the option `access` enabled. Access-enabled databases require docs from regular users to have a new `"_access": ["owner"]` field, that matches the submitting `userCtx.name`. As a result, a user will get full `_all_docs` and `_changes` (and all the other APIs) scoped to just the docs where their name is in the `access` field. See https://github.com/apache/couchdb-documentation/blob/rfc/010-per-document-access/rfcs/010-per-document-access-control.md for some more info on how it all fits together.

There are multiple parts to all this:

1. A new access query server that creates two new indexes `by-access-id` and `by-access-seq` which include all docs in their respective sort orders, but bucketed by user. This required teaching `couch_index` to index deleted docs conditionally.

2. A switch in the mrview code that access `_all_docs` and `_changes` that uses the new indexes for users that are not db or server admins.

3. All the plumbing to handle access information everywhere we need it. It is stored in `#doc` `#doc_info` and `#full_doc_info` records, so we can avoid loading full doc bodies on doc updates.

4. All the actual access validation for document CRUD, especially updates, including replicated updates (mainly in couch_db.erl and couch_db_updater.erl).

5. Teaching the replicator to write `_local` doc checkpoints as the user that owns the replication.

6. All users from the `_users` db now have an implicit role `_users` that will become handy later.

7. Some odds and ends across various modules to make the acquainted with the new reality.

8. The test suite is quite extensive (1k + LOC) and covers most scenarios.

9. Current caveats:
- performance regression in `couchdb_update_conflicts_test.erl` something when from `O(log n)` to `O(n)` or worse. I hope more sets of eyes help with this 👀👀👀.
- a few tests needed adjustments in BC-breaking ways (in terms of sort order IIRC, this is a missing, or superfluous `lists:reverse()` somewhere, that shouldn’t be hard to spot.
- there are a few unused variables, function args, or full functions, that will get cleaned up of course.

## Testing recommendations

`make check` ;)

But also, read the test suite and the linked RFC.

## Related Issues or Pull Requests

#1524 

## Checklist

- [x] Code is written and works correctly
- [x] Changes are covered by tests
- [x] Any new configurable parameters are documented in `rel/overlay/etc/default.ini`
- [ ] A PR for documentation changes has been made in https://github.com/apache/couchdb-documentation

### CODE REVIEW

**Comment by ermouth** [comment]
This is great, gratz! 

One question: it’s not clear to me how /_security response will look for buckets having access restrictions, is there a special field for those restrictions? I mean I want to explicitly mark buckets of the kind in Photon, so how can I do that reading /_security endpoint?

RFC states `admins can grant individual users and groups access to a database using the database’s _security object`, no details.

File: src/couch/src/couch_httpd_auth.erl [hunk 1]
```diff
@@ -87,6 +87,13 @@ basic_name_pw(Req) ->
         nil
     end.
 
+extract_roles(UserProps) ->
+    Roles = couch_util:get_value(<<"roles">>, UserProps, []),
+    case lists:member(<<"_admin">>, Roles) of
+        true -> Roles;
+        _ -> Roles ++ [<<"_users">>]
```
Comment (rnewson) [inline src/couch/src/couch_httpd_auth.erl:94]: `[<<"_users">> | Roles]`

File: src/couch/src/couch_httpd_auth.erl [hunk 2]
```diff
@@ -167,7 +174,7 @@ proxy_auth_user(Req) ->
             Roles = case header_value(Req, XHeaderRoles) of
                 undefined -> [];
                 Else ->
-                    [?l2b(R) || R <- string:tokens(Else, ",")]
+                    [?l2b(R) || R <- string:tokens(Else, ",")] ++ [<<"_users">>]
```
Comment (rnewson) [inline src/couch/src/couch_httpd_auth.erl:177]: as above

**Review by rnewson — CHANGES_REQUESTED** [review]
I've taken a pass at this and have enough feedback to submit this first review. It does look like we're making the right sort of internal changes to store the additional data.

The majority of my comments are about the inconsistencies in the code. For example, in some cases you explicitly handle all expected results of a function, and in others you have a generic `_` field for everything else (even in the case where we are switching on a boolean). I expressed my preference to avoid this sloppiness (aka "defensive programming") and to instead be explicit throughout. There is a place to validate the settings the user has made, and that should be the only place where coercion to type occurs (assuming we don't flat-out reject a malformed request, as I think we should).

Finally, I thought we had moved forward on general formatting conventions. Two blank lines between functions, one blank line between clauses of same function, etc. I've worked that way for some time as has almost all the code I've reviewed from others. Perhaps this was never formally adopted by the couchdb project. Is it the first you are hearing of it, Jan?

**Comment by rnewson** [comment]
another note. In some places you use a list comprehension, in others a lists:map. In both cases you're changing each item of a list to superset or subset of its original content, but using two different mechanisms to do so. It wasn't clear if this was intentional or not. The only difference is error handling, in that the list comprehension approach can silently drop items that fail, whereas lists:map would not. Is that why you chose lc vs map or were the choices arbitrary?

**Comment by stemuk** [comment]
@janl Are there any further plans for this pull request?

**Comment by klehmann** [comment]
We are **very** interested in this new feature, but for our use case (mirroring/migrating data from Notes/Domino databases), just checking the `_access` field values against a single `userCtx.name` value is not sufficient. Would be much more powerful if this second value could be a list as well, e.g. that we populate with all name variants, groups and roles that a user has in the web app that is using CouchDB for data storage.

File: src/couch/src/couch_db.erl [hunk 3]
```diff
@@ -284,26 +295,36 @@ delete_doc(Db, Id, Revisions) ->
 open_doc(Db, IdOrDocInfo) ->
     open_doc(Db, IdOrDocInfo, []).
 
-open_doc(Db, Id, Options) ->
+open_doc(Db, Id, Options0) ->
     increment_stat(Db, [couchdb, database_reads]),
+    Options = case has_access_enabled(Db) of
+        true -> Options0 ++ [conflicts];
```
[the remark this answers is not in the document]
Comment (janl) [reply src/couch/src/couch_db.erl:301]: this is so we can deny access to a doc if the two edit branches disagree on who should have access, the doc becomes admin-only then.

File: src/couch/src/couch_db.erl [hunk 4]
```diff
@@ -724,6 +747,73 @@ security_error_type(#user_ctx{name=null}) ->
...
+    case proplists:get_value(conflicts, Meta) of
+        undefined -> % no conflicts
+            case is_read_from_ddoc_cache(Options) andalso is_per_user_ddoc(Doc) of
+                true -> throw({not_found, missing});
+                _False -> validate_access2(Db, Doc)
+            end;
+        _Else -> % only admins can read conflicted docs in _access dbs
+            case is_admin(Db) of
+                true -> ok;
+                _Else2 -> throw({forbidden, <<"document is in conflict">>})
```
[the remark this answers is not in the document]
Comment (janl) [reply src/couch/src/couch_db.erl:772]: I might have an implementation gap here. The intention is: if leaves disagree on _access, an admin must resolve before a user can proceed. But if leaves agree on _access, a user should just normally be able to use this.

File: src/couch/src/couch_db_updater.erl [hunk 5]
```diff
@@ -676,15 +703,83 @@ update_docs_int(Db, DocsList, LocalDocs, MergeConflicts) ->
...
+% check_access(Db, UserCtx, Access) ->
+%     check_access(Db, UserCtx, couch_db:has_access_enabled(Db), Access).
+%
+% check_access(_Db, UserCtx, false, _Access) ->
+%     true;
+
+% at this point, we already validated this Db is access enabled, so do the checks right away.
+check_access(Db, UserCtx, Access) -> couch_db:check_access(Db#db{user_ctx=UserCtx}, Access).
+
+% TODO: looks like we go into validation here unconditionally and only check in
```
[the remark this answers is not in the document]
Comment (janl) [reply src/couch/src/couch_db_updater.erl:725]: fixed in next push, good eye!

**Comment by janl** [comment]
> another note. In some places you use a list comprehension, in others a lists:map. In both cases you're changing each item of a list to superset or subset of its original content, but using two different mechanisms to do so. It wasn't clear if this was intentional or not. The only difference is error handling, in that the list comprehension approach can silently drop items that fail, whereas lists:map would not. Is that why you chose lc vs map or were the choices arbitrary?

I made no conscious decisions but rather picked the variant that was most prominent in the surrounding code. A cursory glance suggests this is fine where employed, but I’ll have another eye on it once the final PR goes up.

**Comment by janl** [comment]
I’m closing this PR in favour of a new cleaner one shortly since our master/main/3.x rejigger has made this one not show correctly here. I resolved all comments here that I have fixed in the future PR. The remaining ones will need to be looked at there again, I’ll add comments where appropriate, since code moved around quite a bit.

File: src/chttpd/src/chttpd_db.erl [hunk 6]
```diff
@@ -906,16 +925,18 @@ view_cb(Msg, Acc) ->
     couch_mrview_http:view_cb(Msg, Acc).
 
 db_doc_req(#httpd{method='DELETE'}=Req, Db, DocId) ->
-    % check for the existence of the doc to handle the 404 case.
-    couch_doc_open(Db, DocId, nil, []),
-    case chttpd:qs_value(Req, "rev") of
+    % fetch the old doc revision, so we can compare access control
+    % in send_update_doc() later.
+    Doc0 = couch_doc_open(Db, DocId, nil, [{user_ctx, Req#httpd.user_ctx}]),
```
[the remark this answers is not in the document]
Comment (janl) [reply src/chttpd/src/chttpd_db.erl:930]: copied to new PR

File: src/couch/src/couch_db.erl (hunk 3 above)
[the remark this answers is not in the document]
Comment (janl) [reply src/couch/src/couch_db.erl:301]: moved to new PR

File: src/couch/src/couch_db.erl [hunk 7]
```diff
@@ -724,6 +747,73 @@ security_error_type(#user_ctx{name=null}) ->
...
+            {_, true} -> true;
+            _ -> false
+            end
+        end
+    end.
+
+check_name(null, _Access) -> true;
+check_name(UserName, Access) ->
+            lists:member(UserName, Access).
+% nicked from couch_db:check_security
```
[the remark this answers is not in the document]
Comment (janl) [reply src/couch/src/couch_db.erl:810]: moved to new PR

**Pull request closed without merging**